### PR TITLE
Adding computation and plotting of trial by trial metrics

### PR DESCRIPTION
**Pull request by alexpiet**
Updated plot_session_scroller to plot metrics
- users can specify which metrics to plot by passing in a list of metrics to plot
- metrics are plotted with or without FIP data
- metrics are randomly colored, we might want to hardcode a color mapping later
- added a marker to plot whether the port is baited

Trial by trial metrics:
- [x] response_rate, fraction of trials with a response
- [x] gocue_reward_rate, fraction of trials with a reward
- [x] response_reward_rate, fraction of trials with a reward computed only on trials with a response
- [x] choose_right_rate, fraction of trials where mouse chose right computed only on trials with a response
- [x] pL, generative reward probability on the left, just added plotting
- [x] pR, generative reward probability on the right, just added plotting

Future work:
- There are many more metrics to add, but I wanted to merge this in so we have a stable ecosystem for metrics. 
- parameters for computing metrics (window duration, minimum number of events, whether to "center" the window) might want to be metric specific

DEMO
```
nwb.df_trials = compute_all_trial_metrics(nwb)
plot_session_scroller(nwb, plot_bouts=True, metrics=['response_rate','choose_right_rate'])
```
Screenshot 2024-11-13 at 12 03 17 PM

Another example with baiting and generative reward probabilities
```
nwb.df_trials = compute_all_trial_metrics(nwb)
nwb.fip_df = create_fib_df(nwb)
plot_session_scroller(nwb, plot_bouts=True, metrics=['pR','pL'])
```
Screenshot 2024-11-13 at 1 38 31 PM

### CODE REVIEW

File: src/aind_dynamic_foraging_basic_analysis/metrics/trial_metrics.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+"""
+
+import numpy as np
+
+# TODO, we might want to make these parameters metric specific
+WIN_DUR = 15
+MIN_EVENTS = 2
+
+
+def compute_all_trial_metrics(nwb):
```
Comment (hanhou) [inline src/aind_dynamic_foraging_basic_analysis/metrics/trial_metrics.py:14]: I have more trial-based metrics [here](https://github.com/AllenNeuralDynamics/aind-foraging-behavior-bonsai-basic/blob/5bd0db90bf9182a969bcb3c52af77c8bb9a36f19/code/process_nwbs.py#L61) that can be migrated to this function.

Most of my metrics are not binary variables, so they may not need smoothing.

**Comment by hanhou** [comment]
Now the trial metrics are plotted in time-based plot. Do you think it makes sense to also add trial metrics to the [trial-based session plot](https://github.com/AllenNeuralDynamics/aind-dynamic-foraging-basic-analysis/blob/main/src/aind_dynamic_foraging_basic_analysis/plot/plot_foraging_session.py) as well?

Also, it would be great to have a plotly version of all these plots, esp time-based plots with FIP. Then it will be very easy to visualize in Streamlit or Jupyter notebook

**Comment by hanhou** [comment]
I'll go ahead and approve this for now.

**Review by hanhou — APPROVED** [review]

File: src/aind_dynamic_foraging_basic_analysis/metrics/trial_metrics.py (hunk 1 above)
Comment (alexpiet) [reply src/aind_dynamic_foraging_basic_analysis/metrics/trial_metrics.py:14]: Tracking them in this issue: https://github.com/AllenNeuralDynamics/aind-dynamic-foraging-basic-analysis/issues/27

**Comment by alexpiet** [comment]
> * See this issue [add instantaneous stats for each session foraging-behavior-browser#2](https://github.com/AllenNeuralDynamics/foraging-behavior-browser/issues/2)

Tracking them in this consolidated issue: https://github.com/AllenNeuralDynamics/aind-dynamic-foraging-basic-analysis/issues/27

**Comment by alexpiet** [comment]
> trial-based session plot

Yes, made a separate issue here: https://github.com/AllenNeuralDynamics/aind-dynamic-foraging-basic-analysis/issues/28

**Pull request merged**